### PR TITLE
Deploy RPM/Deb to Bintray

### PR DESCRIPTION
**Pull request by SeanTAllen**

### CODE REVIEW

**Comment by SeanTAllen** [comment]
I have a trusty linux vm set up, trying to figure out how all the `make deploy` commands work and what it was doing. currently stuck at:

```
rm -rf build/bin
cp build/release/stable build/"pony-stable"--"1"/usr/lib/pony-stable/-"1"/bin
ln -f -s /usr/lib/pony-stable/-"1"/bin/stable build/"pony-stable"--"1"/usr/bin/stable
cp -r packages build/"pony-stable"--"1"/usr/lib/pony/-"1"/
cp: cannot stat ‘packages’: No such file or directory
make: *** [deploy] Error 1

```

**Comment by SeanTAllen** [comment]
O right, `packages` is the pony packages. Not needed for this.

**Comment by SeanTAllen** [comment]
Now stuck on

```
rm -rf build/bin
cp build/release/stable build/"pony-stable"--"1"/usr/lib/pony-stable/-"1"/bin
ln -f -s /usr/lib/pony-stable/-"1"/bin/stable build/"pony-stable"--"1"/usr/bin/stable
fpm -s dir -t deb -C build/"pony-stable"--"1" -p build/bin --name "pony-stable" --conflicts "pony-stable-master" --conflicts "pony-stable-release" --version  --iteration ""1"" --description "Pony dependency manager" --provides "stable" --provides "pony-stable-release"
All flags should be before the first argument (stray flags found: ["--description", "--provides", "--provides"] {:level=>:warn}
Invalid package configuration: Cannot package the path 'build/pony-stable--1/1', does it exist? {:level=>:error}
make: *** [deploy] Error 1
```

**Comment by SeanTAllen** [comment]
Ok I got `make deploy` to run to completion. i have no idea how to test this. figuring that out.

**Comment by SeanTAllen** [comment]
@killerswan does this look like correct output to you:

---

rm -rf build/bin
cp build/release/stable build/"pony-stable"-0.0.0-f1b631b-"1"/usr/lib/pony-stable/0.0.0-f1b631b-"1"/bin
ln -f -s /usr/lib/pony-stable/0.0.0-f1b631b-"1"/bin/stable build/"pony-stable"-0.0.0-f1b631b-"1"/usr/bin/stable
fpm -s dir -t deb -C build/"pony-stable"-0.0.0-f1b631b-"1" -p build/bin --name "pony-stable" --conflicts "pony-stable-master" --conflicts "pony-stable-release" --version 0.0.0-f1b631b --iteration ""1"" --description "Pony dependency manager" --provides "stable" --provides "pony-stable-release"
Debian packaging tools generally labels all files in /etc as config files, as mandated by policy, so fpm defaults to this behavior for deb packages. You can disable this default behavior with --deb-no-default-config-files flag {:level=>:warn}
Created package {:path=>"build/bin/pony-stable_0.0.0-f1b631b-1_amd64.deb"}
fpm -s dir -t rpm -C build/"pony-stable"-0.0.0-f1b631b-"1" -p build/bin --name "pony-stable" --conflicts "pony-stable-master" --conflicts "pony-stable-release" --version 0.0.0-f1b631b --iteration ""1"" --description "Pony dependency manager" --provides "stable" --provides "pony-stable-release"
Package version '0.0.0-f1b631b' includes dashes, converting to underscores {:level=>:warn}
Created package {:path=>"build/bin/pony-stable-0.0.0_f1b631b-1.x86_64.rpm"}
git archive HEAD > build/bin/"pony-stable"-0.0.0-f1b631b-"1".tar
bzip2 build/bin/"pony-stable"-0.0.0-f1b631b-"1".tar
rm -rf build/"pony-stable"-0.0.0-f1b631b-"1" build/bin/"pony-stable"-0.0.0-f1b631b-"1".tar

**Comment by SeanTAllen** [comment]
I was able to test the debian package. It works. I'm not sure if all the `fpm` options are correct.

Also, it installs in `/usr/bin/stable` rather than `/usr/local/bin/stable`

and 

`/usr/lib/pony-stable` rather than `/usr/local/lib/pony-stable`

Is it customary for RPM/DEB packages to install to `/usr` or `/usr/local`?

I'd assume `/usr/local`.

@jemc thoughts on where this should install? `usr` or `usr/local`?

**Comment by killerswan** [comment]
Typically, `/usr/local` is where stuff you build yourself locally should go.  These are not that. :)

**Comment by killerswan** [comment]
Are you sure you want this to be version 0.0.0?

**Comment by killerswan** [comment]
I would look at the makefile and simply this bit as shown in the FPM calls above:
```
--conflicts "pony-stable-master" --conflicts "pony-stable-release" --version 0.0.0-f1b631b --iteration ""1"" --description "Pony dependency manager" --provides "stable" --provides "pony-stable-release"
```

Probably, it doesn't conflict with anything, and it provides "pony-stable".  Everything else is cruft from `ponyc` where we had previously shipped packages that called themselves `...-release` and `...-master`.

File: Makefile [hunk 1]
```diff
@@ -35,4 +37,41 @@ all: $(binary)
...
+	$(SILENT)rm -rf build/bin
+	@mkdir -p build/bin
+	@mkdir -p $(package)/usr/bin
+	@mkdir -p $(package)/usr/lib/pony-stable/$(package_version)/bin
+
+	$(SILENT)cp $(BUILD_DIR)/stable $(package)/usr/lib/pony-stable/$(package_version)/bin
+
+	$(SILENT)ln -f -s /usr/lib/pony-stable/$(package_version)/bin/stable $(package)/usr/bin/stable
+	$(SILENT)fpm -s dir -t deb -C $(package) -p build/bin --name $(package_name) --conflicts "pony-stable-master" --conflicts "pony-stable-release" --version $(package_base_version) --iteration "$(package_iteration)" --description "Pony dependency manager" --provides "stable" --provides "pony-stable-release"
+	$(SILENT)fpm -s dir -t rpm -C $(package) -p build/bin --name $(package_name) --conflicts "pony-stable-master" --conflicts "pony-stable-release" --version $(package_base_version) --iteration "$(package_iteration)" --description "Pony dependency manager" --provides "stable" --provides "pony-stable-release"
```
Comment (killerswan) [inline Makefile]: Here, particularly: unless these conflicts are packages that exist somewhere, we don't need to list them in the new packages.  So take out these `--conficts`.  And take out the `--provides "pony-stable-release"`, too.

**Review by killerswan — COMMENTED** [review]
As I've described, I think the RPM and DEB packages don't need --conficts and as many --provides, but once those are cleaned up it looks good to me.

Caveat: I have pneumonia and haven't been in this codebase recently enough anyways, so there may be something I'm not spotting.  But it all looks very familiar, just like what we set up with ponyc.

If you wanna test it out end to end, you can use another Bintray and Travis account pair!

**Comment by SeanTAllen** [comment]
@killerswan why would this be version 0.0.0? there's a separate PR that updated version (and release scripts ala Pony that will update automatically). is there anything else related to 0.0.0 as version that I missed?

**Comment by SeanTAllen** [comment]
@killerswan i made updates, how does this look?

one things i am unclear on.

the command installed is `stable`. should the `--provides` be `pony-stable`. my guess is, yes, i should change it to be `--provides="pony-stable"`

**Comment by killerswan** [comment]
As far as what it provides, that makes sense to me, or `--provides $(package_name)`.

Otherwise LGTM!

**Comment by SeanTAllen** [comment]
Thanks @killerswan

**Pull request merged**